pkg/asset/asset.go

```diff
@@ -118,14 +118,96 @@ type Config struct {
+	ServiceCIDRs           []*net.IPNet
+	APIServiceIPs          []net.IP
+	DNSServiceIPs          []net.IP
 	CloudProvider          string
 	NetworkProvider        string
 	BootstrapSecretsSubdir string
 	Images                 ImageVersions
+
+	// PodCIDR describes the networking subnet to be used for inter-pod networking.
+	//
+	// Deprecated: PodCIDR exists only for compatibility with older external
+	// systems.  Please use PodCIDRs instead, which allows for dual-stack
+	// configurations.
+	PodCIDR *net.IPNet
+
+	// ServiceCIDR describes the networking subnet to be used to expose services.
+	//
+	// Deprecated: ServiceCIDR exists only for compatibility with older external
+	// systems.  Please use ServiceCIDRs instead, which allows for dual-stack
+	// configurations.  If both are specified, only ServiceCIDRs will be used.
+	ServiceCIDR *net.IPNet
+
+	// APIServiceIP describes the in-cluster IP address by which the API Servers may be reached.
+	//
+	// Deprecated: APIServiceIP exists only for compatibility with older
+	// external systems.  Please use APIServiceIPs instead, which allows for
+	// dual-stack configurations.  If both are specified, only APIServiceIPs
+	// will be used.
+	APIServiceIP net.IP
+
+	// DNSServiceIP describes the in-cluster IP address by which the cluster DNS servers may be reached.
+	//
+	// Deprecated:  DNSServiceIP exists only for compatibility with older
+	// external systems.  Please use DNSServiceIPs instead, which allows for
+	// dual-stack configurations.  If both are specified, only DNSServiceIPs
+	// will be used.
+	DNSServiceIP net.IP
+}
+
+// BindAllAddress indicates the address to use when binding all IPs.  If this
+// is an IPv6 or dual-stack system, `::` will be returned.  Otherwise
+// `0.0.0.0`.
+func (c *Config) BindAllAddress() string {
+	for _, ip := range c.APIServiceIPs {
+		if ip == nil || ip.IsLoopback() || ip.IsUnspecified() {
+			continue
+		}
+
+		if ip.To4() == nil && ip.To16() != nil {
+			return "::"
+		}
+	}
+	return "0.0.0.0"
+}
+
+// ServiceCIDRsString returns a "," concatenated string for the ServiceCIDRs
+func (c *Config) ServiceCIDRsString() string {
+	return joinStringsFromSliceOrSingle(stringerSlice(c.ServiceCIDRs), c.ServiceCIDR.String())
+}
+
+// PodCIDRsString returns a "," concatenated string for the PodCIDRs
+func (c *Config) PodCIDRsString() string {
+	return joinStringsFromSliceOrSingle(stringerSlice(c.PodCIDRs), c.PodCIDR.String())
+}
+
+// APIServiceIPsString returns a "," concatenated string for the APIServiceIPs
+func (c *Config) APIServiceIPsString() string {
+	return joinStringsFromSliceOrSingle(stringerSlice(c.APIServiceIPs), c.APIServiceIP.String())
+}
+
+// DNSServiceIPsString returns a "," concatenated string for the DNSServiceIPs
+func (c *Config) DNSServiceIPsString() string {
+	return joinStringsFromSliceOrSingle(stringerSlice(c.DNSServiceIPs), c.DNSServiceIP.String())
+}
+
+func stringerSlice(in interface{}) []string {
+	r := reflect.ValueOf(in)
```

Comment:
I avoid `reflect` like the plague, but I don't think there is a better way to keep this generic.

Reply:
That was exactly the conundrum.  On the one hand, it wouldn't kill us to repeat the code; on the other, using `reflect` always has a bit of code smell.  In my opinion (obviously, I guess), this was a safe enough usage which should not be performance-affecting... but I'm willing to do it either way, depending on general consensus.

Reply:
I'm fine as is ... This is not in the critical path so the hit is basically nothing.

Reply:
Ok